rust/src/main.rs

```diff
@@ -50,65 +50,48 @@ where
-        .to_u32()
-        .ok_or("Number of documents too large (must fit in u32)")?;
-
-    let input = BufReader::new(&input_file);
-    let mut sizes = BufWriter::new(File::create(format!("{}.sizes", output_basename))?);
-    let mut trecids = BufWriter::new(File::create(format!("{}.documents", output_basename))?);
-
-    let bar = ProgressBar::new(num_documents as u64);
-    bar.set_style(pb_style());
-    bar.set_draw_delta(num_documents as u64 / 100);
-    sizes.write_all(&num_documents.to_ne_bytes())?;
-    for (expected_docid, line) in input.lines().enumerate() {
-        let line = line?;
-        let mut columns = line.split('\t');
-        let docid: u32 = columns.next().ok_or("Corrupted file")?.parse()?;
-        let trecid = columns.next().ok_or("Corrupted file")?;
-        let length: u32 = columns.next().ok_or("Corrupted file")?.parse()?;
-        assert_eq!(
-            docid as usize, expected_docid,
-            "Document sizes must come in order"
-        );
-        sizes.write_all(&length.to_ne_bytes())?;
-        writeln!(trecids, "{}", trecid)?;
-        bar.inc(1);
-    }
-    bar.finish();
-    Ok(num_documents)
+fn print_header(hdr: Header) {
+    println!("----- CIFF HEADER -----");
+    println!("Version: {}", hdr.get_version());
+    println!("No. Postings Lists: {}", hdr.get_num_postings_lists());
+    println!("Total Postings Lists: {}", hdr.get_total_postings_lists());
+    println!("No. Documents: {}", hdr.get_num_docs());
+    println!("Total Documents: {}", hdr.get_total_docs());
+    println!("Total Terms in Collection {}", hdr.get_total_terms_in_collection());
+    println!("Average Document Length: {}", hdr.get_average_doclength());
+    println!("Description: {}", hdr.get_description());
+    println!("-----------------------");
 }
 
 fn gen(args: Args) -> Result<()> {
-    let mut postings_reader = File::open(args.postings)?;
-    let mut input = CodedInputStream::new(&mut postings_reader);
+    let mut ciff_reader = File::open(args.ciff_file)?;
+    let mut input = CodedInputStream::new(&mut ciff_reader);
     let mut documents = BufWriter::new(File::create(format!("{}.docs", args.output))?);
     let mut frequencies = BufWriter::new(File::create(format!("{}.freqs", args.output))?);
     let mut terms = BufWriter::new(File::create(format!("{}.terms", args.output))?);
 
-    let num_documents = write_document_lengths(&args.document_lengths, &args.output)?;
+    // Read protobuf header
+    let header = input.read_message::<Header>()?;
+    let num_documents = header.get_num_docs();
+    print_header(header.clone());
+
 
     eprintln!("Processing postings...");
     encode_sequence(&mut documents, 1, [0_u32].iter().copied())?;
     let bar = ProgressBar::new(262);
     bar.set_style(pb_style());
     bar.set_draw_delta(10);
-    while !input.eof()? {
+    let mut postings_seen = 0;
+    let expected_postings = header.get_num_postings_lists();
+    while postings_seen < expected_postings {
```

Comment:
The rule of thumb in Rust is to avoid mutable variables if possible. Here, you can do that by replacing `while` with `for`:
```rust
for _ in 0..header.get_num_postings_lists() {
```

Note that `_` indicates that this variable will not be used (similar to python).